Include/internal/pycore_runtime.h

```diff
@@ -108,6 +143,7 @@ typedef struct _Py_DebugOffsets {
         uint64_t instr_ptr;
         uint64_t localsplus;
         uint64_t owner;
+        uint64_t stackpointer;
```

Comment:
This is needed because there is no other way to get the current running coroutine object from the frame with the current new layout from external profilers they can only read memory. Please check the tests. 

If you have a better alternative I am happy to move to that but right now I don't see any other possibility

Reply:
> The stack pointer is not always readable, nor is it always correct, nor is it part of the ABI, and when it is correct or not may change at a whim.

Can you point me at (pun not intended) a situation where the stack pointer will be invalid in a way that will make the external introspection (done in the tests here) not work?

> In fact, it didn't even exist in 3.13, and may not exist again in future versions.

Is there a design that you have in mind that would result in the removal of this entirely?

AFAICT if we decided to drop deferred refcounting, we'd return to what `stacktop` was doing in the past, which was more-less equivalent for our purpose here. Moreover, @Fidget-Spinner claims the stack pointer is a better design regardless of deferred refcounting.

Reply:
> In fact, it didn't even exist in 3.13, and may not exist again in future versions.

@markshannon Mark, I think the contract here is that CPython can move freely with its refactorings and changes. If `stackpointer` breaks or becomes completely non-meaningful we'll just mark our test "xfail" and figure another way to introspect the particular future version of CPython. I wouldn't worry about these offsets at all, we're not saying that they are public API.

Same as internal layout of Set/Dict/asyncio objects -- they are all subject to break from one version to another. 

I really think this particular change is fine -- CPython doesn't care nor does guarantee anything whatsoever here.

Reply:
> Is there a design that you have in mind that would result in the removal of this entirely?

We experimented with a register VM. We ultimately decided against it, but if we had gone for it there would have been no stack pointer.

>  If stackpointer breaks or becomes completely non-meaningful we'll just mark our test "xfail" 

The problem is that optimizations may leave the tests working, but the stack pointer could easily be meaningless in other places.
I would strongly recommend that you avoid relying on the stack pointer in any way.

We expect that the stack will consist of zero or more values in registers followed by the remaining values in memory,
with the stack pointer pointing to some offset from the in-memory stack top.
The offset and number of values in registers will be not be knowable at runtime.

Reply:
> Please check the tests.

@pablogsal 
Where? I don't see where it is used.

Reply:
> > Please check the tests.
> 
> @pablogsal Where? I don't see where it is used.

https://github.com/python/cpython/pull/124640/files#diff-08adf6717de7c662b4e6960c5c428e6c46b1ef5370b562a4b0c92ecb071ce733R728

Reply:
I will copy It here for your convenience:

```
      uintptr_t stackpointer_addr;
         err = read_py_ptr(
             pid,
             coro_address + offsets->gen_object.gi_iframe +
                 offsets->interpreter_frame.stackpointer,
             &stackpointer_addr);
         if (err) {
             return -1;
         }

         if ((void*)stackpointer_addr != NULL) {
             uintptr_t gi_await_addr;
             err = read_py_ptr(
                 pid,
                 stackpointer_addr - sizeof(void*),
                 &gi_await_addr);
             if (err) {
                 return -1;
             }
```

as you can see is needed to go from the coroutine to the `gi_await` field.